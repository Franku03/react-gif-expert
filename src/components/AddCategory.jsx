import { useState } from "react"
import PropTypes from "prop-types";

export const AddCategory = ({ onNewCategory }) => {

    const [inputValue, setInputValue] = useState('');

    const onInputChange = ({ target }) => {
        // let's remember that in the function calling, we are passing it the event object when we do a calling without arguments
        setInputValue(target.value );
    }

    const onSubmit = (event) =>{
        event.preventDefault();
        if( inputValue.trim().length <= 1) return;

        // setCategories( categories => [ inputValue, ...categories ]);
        onNewCategory( inputValue.trim() );
        setInputValue('');
    }

    return (
       
        <form onSubmit={ onSubmit } aria-label="form">
            <input 
                type="text"
                placeholder="Buscar Gifs"
                value={ inputValue }
                onChange={ onInputChange }
            />
        </form>

    )
}



AddCategory.propTypes = {
    onNewCategory: PropTypes.func.isRequired,
}