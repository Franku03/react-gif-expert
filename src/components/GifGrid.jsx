import PropTypes from "prop-types";
import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";


export const GifGrid = ({ category }) => {

    const { images, isLoading } = useFetchGifs( category );

    return (
        <>
            <h3>{ category }</h3>

            {
                isLoading && ( <h2>Cargando...</h2> )
            }


            <div className="card-grid">
                {
                    images.map( ( img ) => (
                        <GifItem 
                            key={ img.id } 
                            { ...img }
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}




/*
    ({ category }) => {
        ? Nos permite disparar efectos secundarios
        Al dejar las dependencias (2do paramétro) vacias, significa que el hook solo se dispara la primera vez que se renderiza el componente
        ¡ Also, useEffect NO permite pasarle una función async, porque debe recibir un callback, no una promesa, que es lo que nos regresa una función async
        Si queremos hacer algo asíncrono debemos usar sintaxis de promises, o llamar a una función asíncrona externa desde el useEffect

        useEffect( () => {
            getGifs( category );
        }, [ ]);

        ! Mala práctica: llamar funciones desde un functional component solo hará que se llamen cada vez que se vuelva a renderizar el componente
        getGifs( category );
    }
*/
