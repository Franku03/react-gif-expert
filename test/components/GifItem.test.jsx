import { render, screen } from "@testing-library/react";
import { GifItem } from "../../src/components/GifItem";

describe('Pruebas en <GifItem />', () => {

    const image = {
        id: 1,
        title: 'Gintoki',
        url: 'https://gintama.com/gintoki.png',
    }

    test('Debe hacer match con el snapshot', () => {
        const { container } = render( <GifItem key={ image.id } { ...image } /> );
        expect( container ).toMatchSnapshot();
    });

    test('Debe de mostrar la imagen con el URL y el ALT indicado', () => {
        render(<GifItem key={ image.id } { ...image }  />);
        // expect( screen.getByRole('img').src ).toBe(  image.url );

        const { src, alt} = screen.getByRole('img');
        expect( src ).toBe( image.url );
        expect( alt ).toBe( image.title );

    })

    test('Debe de mostrar el título en el componente', () => {
        render( <GifItem key={ image.id } { ...image }  /> );
        expect( screen.getByText( image.title ) ).toBeTruthy();
        // El title del <img/> element no hace conflicto con esta petición de getByText porque en ese elemento title es un atributo, no un texto
    });

});