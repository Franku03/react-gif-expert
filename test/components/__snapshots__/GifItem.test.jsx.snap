// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <GifItem /> Debe hacer match con el snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Gintoki"
      src="https://gintama.com/gintoki.png"
    />
    <p>
      Gintoki
    </p>
  </div>
</div>
`;
